sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -966,6 +952,27 @@ object ReplaceDistinctWithAggregate extends Rule[LogicalPlan] {
 }
 
 /**
+ * Replaces logical [[Intersect]] operator with a left-semi [[Join]] operator.
+ * {{{
+ *   SELECT a1, a2 FROM Tab1 INTERSECT SELECT b1, b2 FROM Tab2
+ *   ==>  SELECT a1, a2 FROM Tab1, Tab2 ON a1<=>b1 AND a2<=>b2
+ * }}}
+ */
+object ReplaceIntersectWithLeftSemi extends Rule[LogicalPlan] {
+  private def buildCond (left: LogicalPlan, right: LogicalPlan): Seq[Expression] = {
+    require(left.output.length == right.output.length)
```

Comment:
Yeah, it has a checking logic in analyzer:
https://github.com/apache/spark/blob/d084a2de3271fd8b0d29ee67e1e214e8b9d96d6d/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala#L186-L190